zebra-chain/src/history_tree.rs

```
@@ -0,0 +1,245 @@
+    peaks: BTreeMap<u32, Entry>,
+    /// The height of the most recent block added to the tree.
+    current_height: Height,
+}
+
+impl HistoryTree {
+    /// Create a new history tree with a single block.
+    pub fn new_from_block(
+        network: Network,
+        block: Arc<Block>,
+        sapling_root: &sapling::tree::Root,
+        _orchard_root: Option<&orchard::tree::Root>,
+    ) -> Result<Self, io::Error> {
+        let height = block
+            .coinbase_height()
+            .expect("block must have coinbase height during contextual verification");
+        let network_upgrade = NetworkUpgrade::current(network, height);
+        // TODO: handle Orchard root
+        let (tree, entry) = Tree::new_from_block(network, block, sapling_root)?;
+        let mut peaks = BTreeMap::new();
+        peaks.insert(0u32, entry);
+        Ok(HistoryTree {
+            network,
+            network_upgrade,
+            inner: tree,
+            size: 1,
+            peaks,
+            current_height: height,
+        })
+    }
+
+    /// Add block data to the tree.
+    ///
+    /// # Panics
+    ///
+    /// If the block height is not one more than the previously pushed block.
+    pub fn push(
+        &mut self,
+        block: Arc<Block>,
+        sapling_root: &sapling::tree::Root,
+        _orchard_root: Option<&orchard::tree::Root>,
+    ) -> Result<(), HistoryTreeError> {
+        // Check if the block has the expected height.
+        // librustzcash assumes the heights are correct and corrupts the tree if they are wrong,
+        // resulting in a confusing error, which we prevent here.
+        let height = block
+            .coinbase_height()
+            .expect("block must have coinbase height during contextual verification");
+        if height - self.current_height != 1 {
+            panic!(
+                "added block with height {:?} but it must be {:?}+1",
+                height, self.current_height
+            );
+        }
```

Should we panic, or should we return an `Err` ?

---

Good question! When this gets called, the first part of the contextual verification should have already been called, which ensures that the height is correct. I'd argue that passing a block the incorrect height would be a programming error and not something that could come up in the regular working of Zebra, so a panic would suffice.

However it's a simple change to make it an error. What do you think?

---

Here are some guidelines:

Would continuing on error risk corrupting the memory or disk state, accepting invalid data, or losing user funds? Then we should panic.

Also if it's a programmer bug we should panic.

If it's validating arbitrary network data we should return an error.

Otherwise it's a judgement call. (But prefer errors.)

---

I'm good panicking here then, hopefully it never triggers 😁